src-tauri/src/lib.rs

```diff
@@ -0,0 +1,153 @@
+use headless_chrome::browser::tab;
+use headless_chrome::protocol::cdp::Page;
+use headless_chrome::protocol::cdp::Target::CreateTarget;
+use headless_chrome::{browser, Browser, LaunchOptions};
+use regex::Regex;
+use std::sync::Arc;
+use std::thread;
+use std::time::Duration;
+use tokio::sync::Mutex;
+use totp_rs::{Algorithm, Secret, TOTP};
+
+// Login with TOTP
+// Using the chromiumoxide library to handle the login process for the website
+pub async fn request_token(
+    browser: Browser,
+) -> Result<(String, String), Box<dyn std::error::Error>> {
+    const LOGIN_URL: &str =
+        "http://ec2-3-110-151-1.ap-south-1.compute.amazonaws.com/tmp/simulate/login.html";
+    const WAITING_TIME: u64 = 3; //s
+    let new_tab = browser.new_tab()?;
+    let navigated_tab = new_tab.navigate_to(LOGIN_URL)?.wait_until_navigated()?;
+
+    // Wait for the page to load
+    // Step 2: Enter username and password
+    navigated_tab
+        .find_element("input#userid")?
+        .type_into("AB0101")?;
+
+    navigated_tab
+        .find_element("input#password")?
+        .type_into("R@ndOmP@$$wOrd123!")?;
+
+    navigated_tab.find_element("button")?.click()?;
+
+    // Step 3: Generate TOTP token
+    let totp = TOTP::new(
+        Algorithm::SHA1,
+        6,
+        1,
+        30,
+        Secret::Raw("TestSecretSuperSecret".as_bytes().to_vec())
+            .to_bytes()
+            .unwrap(),
+    )
+    .unwrap();
+    let token = totp.generate_current().unwrap();
+
+    // Step 4: Enter TOTP token
+    navigated_tab.wait_for_element("input#userid");
+    navigated_tab
+        .find_element("input#userid")?
+        .type_into(token.as_str())?;
+
+    // Submit the form
+    navigated_tab
+        .find_element("button.button-orange.wide")?
+        .click()?;
+    navigated_tab.wait_for_element("strong#request_token");
+
+    let requested_token = navigated_tab
+        .find_element("strong#request_token")?
+        .get_inner_text()?;
```

Comment:
We cannot use this mechanism to get the request token since I'm just printing it in the DOM.

The correct way to get the request token is from the URL. The actual web app will do a callback with the request token URL param and it needs to be pulled from there, not from the DOM.

Reply:
✅